File: 40_APP/PVM/Core/Classifier.swift
import Foundation

/// THE CLASSIFIER — where a photo belongs, how sure we are, why, and stopping as early
/// as the evidence allows. Mirrors `30_ENGINE/pvm/classifier.py`.
///
/// The escalation policy is L1-B made executable: run the cheapest tier; if the answer
/// is already settled — a leaf, above the settle threshold — stop; otherwise buy the
/// next tier and ask again.
///
/// Two consequences, both features:
///  * Most of a real camera roll settles at metadata. A photo with a GPS fix and a date
///    is `Places > United Kingdom > London` and `Timeline > 2025` for free.
///  * When the evidence supports "Screenshots" and nothing narrower, the answer is
///    `Screenshots`. It does not guess a leaf. A library full of confidently wrong
///    leaves is worse than one that admits the depth it earned, because the user cannot
///    tell the two apart until they go looking for something.
public struct Classifier {

    private let context: LibraryContext
    private let budget: Tier
    private static var regexCache: [String: NSRegularExpression] = [:]
    private static let cacheLock = NSLock()

    public init(context: LibraryContext, budget: Tier = .text) {
        self.context = context
        self.budget = budget
    }

    public func classify(_ a: AssetSignals) -> Classification {
        var c = Classification(assetID: a.assetID)
        c.tiersSpent.insert(.metadata)

        timeline(a, &c)
        metadataRoot(a, &c)

        // Screenshots are exempt from the visual tiers. Face clustering and scene
        // classification on a chat window produce noise at full price — the only signal
        // that tells you anything about a screenshot is its text.
        let visualWorthIt = !(a.isScreenshot || a.isScreenRecording)

        if visualWorthIt, !settled(c), budget >= .visual {
            c.tiersSpent.insert(.visual)
            scenes(a, &c)
        }
        if visualWorthIt, !settled(c), budget >= .faces {
            c.tiersSpent.insert(.faces)
            faces(a, &c)
        }
        if budget >= .text, wantsText(a, c) {
            c.tiersSpent.insert(.text)
            text(a, &c)
        }

        // Places runs LAST among the content roots, and that ordering is load-bearing.
        // Run first, a GPS fix settled every photographed document as `Places > … >
        // London` at 0.91, the escalation policy saw a settled leaf, and the OCR pass
        // that would have recognised the passport never happened. Nine of nine
        // foreground documents were lost that way, silently, with a plausible answer in
        // their place. Location is context; it is not what the thing IS.
        places(a, &c)
        travel(a, &c)
        pruneImpliedAncestors(&c)
        crossList(&c)
        unfiled(&c)
        return c
    }

    // MARK: - escalation policy

    private func settled(_ c: Classification) -> Bool {
        guard let p = c.primary else { return false }
        return TaxonomyRuntime.isLeaf(p.path) && p.confidence >= Rules.settleAt
    }

    /// Whether to READ text that exists. Whether to SPEND an OCR pass is the C-1 gate,
    /// and it lives upstream where the pixels are. Duplicating that decision here on
    /// metadata the classifier cannot verify would be two gates disagreeing about one
    /// photo. Screenshots always read their text: the subtype settles the root, and
    /// only the text can ever open the leaf.
    private func wantsText(_ a: AssetSignals, _ c: Classification) -> Bool {
        guard a.ocrRan, !a.ocrText.isEmpty else { return false }
        if let p = c.primary, TaxonomyRuntime.root(of: p.path) == "Screenshots" { return true }
        return !settled(c)
    }

    // MARK: - tiers

    private func timeline(_ a: AssetSignals, _ c: inout Classification) {
        guard let when = a.createdAt, let year = a.year else {
            c.notes.append("no capture date — this asset cannot be placed on the timeline")
            return
        }
        guard let path = TaxonomyRuntime.ensure("Timeline" + Taxonomy.separator + String(year))
        else { return }
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_GB")
        f.dateFormat = "d MMMM yyyy"
        guard let e = Evidence(signal: "createdAt", tier: .metadata, weight: 0.98,
                               reason: "taken on \(f.string(from: when))"),
              let assignment = Assignment(path: path, evidence: [e]) else { return }
        c.add(assignment)
    }

    private func metadataRoot(_ a: AssetSignals, _ c: inout Classification) {
        if a.isVideo {
            // The tree has eleven roots and none is Video. A real gap (FC-2), recorded
            // rather than papered over by filing video as a photo.
            c.notes.append("video: the taxonomy has no node for video assets (FC-2)")
        }
        if a.isScreenshot || a.isScreenRecording {
            if let e = Evidence(signal: "isScreenshot", tier: .metadata, weight: 0.95,
                                reason: "the system recorded this as a screenshot"),
               let assignment = Assignment(path: "Screenshots", evidence: [e], isPrimary: true) {
                c.add(assignment)
            }
            return
        }
        if a.source == "downloaded" {
            if let e = Evidence(signal: "source", tier: .metadata, weight: 0.88,
                                reason: "saved from another app rather than taken with the camera"),
               let assignment = Assignment(path: "Downloads", evidence: [e], isPrimary: true) {
                c.add(assignment)
            }
        }
    }

    /// Paperwork is not a place memory. A receipt photographed at the kitchen table does
    /// not belong on the Places shelf next to the holiday photos.
    private static let notPlaces: Set<String> = ["Documents", "Purchases", "Screenshots", "Downloads"]

    private func places(_ a: AssetSignals, _ c: inout Classification) {
        guard let geo = a.geo, geo.source == .exif else { return }
        guard let place = a.place, let country = place.country else { return }
        if c.assignments.contains(where: { Classifier.notPlaces.contains(TaxonomyRuntime.root(of: $0.path)) }) {
            return
        }
        var parts = ["Places", country]
        var reason = "the location saved with the photo is in \(country)"
        if let city = place.city {
            parts.append(city)
            reason = "the location saved with the photo is in \(city), \(country)"
        }
        guard let path = TaxonomyRuntime.ensure(parts.joined(separator: Taxonomy.separator)) else { return }
        let weight = min(0.92, 0.55 + 0.40 * place.confidence) * geo.confidence
        let isPrimary = !c.assignments.contains { $0.isPrimary }
        if let e = Evidence(signal: "geo", tier: .metadata, weight: weight, reason: reason),
           let assignment = Assignment(path: path, evidence: [e], isPrimary: isPrimary) {
            c.add(assignment)
        }
    }

    /// Travel is a library-level judgement, not a per-photo one: a run of days spent
    /// away from home. One photo in Tokyo is not a trip.
    private func travel(_ a: AssetSignals, _ c: inout Classification) {
        guard let trip = context.trip(for: a.createdAt, geo: a.geo) else { return }
        guard let path = TaxonomyRuntime.ensure("Travel" + Taxonomy.separator + trip.label) else { return }
        let cal = Calendar(identifier: .gregorian)
        let span = (cal.dateComponents([.day], from: trip.start, to: trip.end).day ?? 0) + 1
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_GB")
        f.dateFormat = "MMMM yyyy"
        let where_ = trip.city ?? trip.country ?? "another place"
        if let e = Evidence(signal: "geo+createdAt", tier: .metadata, weight: 0.80,
                            reason: "part of \(span) days away from home in \(where_), \(f.string(from: trip.start))"),
           let assignment = Assignment(path: path, evidence: [e]) {
            c.add(assignment)
        }
    }

    private func faces(_ a: AssetSignals, _ c: inout Classification) {
        let named = a.namedPeople
        if named.isEmpty {
            if !a.faceClusters.isEmpty {
                // An unnamed cluster still changes the risk class, which is the point:
                // "someone is in this photo" is enough to stop treating it as disposable.
                c.notes.append("unnamed person present")
            }
            return
        }
        let path: String?
        let reason: String
        let weight: Double
        if named.count == 1 {
            path = TaxonomyRuntime.ensure("People" + Taxonomy.separator + named[0])
            reason = "\(named[0]) was recognised in this photo"
            weight = 0.88
        } else {
            path = "People > Groups"
            reason = "recognised " + named.joined(separator: ", ") + " together in this photo"
            weight = 0.85
        }
        guard let p = path else { return }
        let isPrimary = !c.assignments.contains { $0.isPrimary }
        if let e = Evidence(signal: "faceClusters", tier: .faces, weight: weight, reason: reason),
           let assignment = Assignment(path: p, evidence: [e], isPrimary: isPrimary) {
            c.add(assignment)
        }
    }

    private func scenes(_ a: AssetSignals, _ c: inout Classification) {
        var best: [String: SceneLabel] = [:]
        for label in a.sceneLabels where label.confidence >= Rules.sceneFloor {
            guard let path = Rules.sceneMap[label.identifier] else { continue }
            if let existing = best[path], existing.confidence >= label.confidence { continue }
            best[path] = label
        }
        for (path, label) in best {
            let weight = max(0.30, min(0.86, 0.30 + 0.60 * label.confidence))
            let pretty = label.identifier.replacingOccurrences(of: "_", with: " ")
            let root = TaxonomyRuntime.root(of: path)
            let isPrimary = !c.assignments.contains { $0.isPrimary }
                && (root == "Objects" || root == "Downloads")
            if let e = Evidence(signal: "sceneLabels", tier: .visual, weight: weight,
                                reason: "a \(pretty) was recognised in the picture"),
               let assignment = Assignment(path: path, evidence: [e], isPrimary: isPrimary) {
                c.add(assignment)
            }
        }
    }

    private static func regex(_ pattern: String) -> NSRegularExpression? {
        cacheLock.lock(); defer { cacheLock.unlock() }
        if let cached = regexCache[pattern] { return cached }
        guard let made = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]) else {
            return nil
        }
        regexCache[pattern] = made
        return made
    }

    private func matches(_ pattern: String, _ text: String) -> Bool {
        guard let re = Classifier.regex(pattern) else { return false }
        let range = NSRange(text.startIndex..<text.endIndex, in: text)
        return re.firstMatch(in: text, options: [], range: range) != nil
    }

    private func text(_ a: AssetSignals, _ c: inout Classification) {
        let body = a.ocrText
        guard !body.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return }
        let root = c.primary.map { TaxonomyRuntime.root(of: $0.path) }

        // Within a group the first match wins, so specific rules precede general ones
        // in the generated table — `Documents > Identity > Passports` before `Documents`.
        let groups = [Rules.screenshotRules, Rules.documentRules,
                      Rules.purchaseRules, Rules.workTextRules]
        for group in groups {
            for rule in group {
                if let required = rule.requiresRoot, required != root { continue }
                guard matches(rule.pattern, body) else { continue }
                var isPrimary = !c.assignments.contains { $0.isPrimary }
                if root == "Screenshots", rule.requiresRoot == "Screenshots" {
                    isPrimary = true
                    for i in c.assignments.indices where c.assignments[i].path == "Screenshots" {
                        c.assignments[i].isPrimary = false
                    }
                }
                if let e = Evidence(signal: "ocrText", tier: .text, weight: rule.weight,
                                    reason: rule.reason),
                   let assignment = Assignment(path: rule.path, evidence: [e], isPrimary: isPrimary) {
                    c.add(assignment)
                }
                break
            }
        }
    }

    /// `Screenshots` and `Screenshots > Temporary > Pickup Codes` are one entry, not
    /// two. The browse tree rolls counts up from the leaves, so leaving both in would
    /// count the asset twice in its own parent. The ancestor's evidence is merged into
    /// the descendant rather than discarded — it is half the reason the leaf is right,
    /// and the user is owed the whole reason.
    private func pruneImpliedAncestors(_ c: inout Classification) {
        let held = Set(c.assignments.map { $0.path })
        for assignment in c.assignments {
            let deeper = held.filter { $0 != assignment.path
                && TaxonomyRuntime.ancestors(of: $0).contains(assignment.path) }
            guard let target = deeper.min(by: { TaxonomyRuntime.depth($0) < TaxonomyRuntime.depth($1) })
            else { continue }
            for i in c.assignments.indices where c.assignments[i].path == target {
                c.assignments[i].evidence.append(contentsOf: assignment.evidence)
                c.assignments[i].isPrimary = c.assignments[i].isPrimary || assignment.isPrimary
            }
            c.assignments.removeAll { $0.path == assignment.path }
        }
    }

    /// `_also_in` from the tree. One original, two entries (§3) — the receipt the user
    /// thinks of as paperwork and the receipt they think of as a purchase are the same
    /// photo, filed once.
    private func crossList(_ c: inout Classification) {
        for assignment in c.assignments {
            guard let also = TaxonomyRuntime.crossListing(of: assignment.path) else { continue }
            if let e = Evidence(signal: "taxonomy", tier: .metadata,
                                weight: min(0.9, assignment.confidence),
                                reason: "also filed under \(also) — the same photo, not a copy "
                                        + "(cross-listed from \(assignment.path))"),
               let extra = Assignment(path: also, evidence: [e],
                                      crossListedFrom: assignment.path) {
                c.add(extra)
            }
        }
    }

    /// An asset with a date and nothing else is on the timeline and nowhere else.
    /// Saying so is more useful than inventing a home for it, and it is the honest
    /// count of how much of a library the engine cannot yet explain.
    private func unfiled(_ c: inout Classification) {
        let content = c.assignments.filter { TaxonomyRuntime.root(of: $0.path) != "Timeline" }
        if content.isEmpty {
            c.notes.append("unfiled: no signal placed this asset anywhere but the timeline")
        }
    }
}

File: 30_ENGINE/pvm/classifier.py
# -*- coding: utf-8 -*-
"""
THE CLASSIFIER.

Given one asset's signals and the library it lives in, decide where it belongs, how
sure we are, and why — then stop as early as the evidence allows.

The escalation policy is the part that matters, and it is L1-B (Information-Change-
First / Minimum Necessary Inference) made executable:

    Run the cheapest tier. If the answer is already settled — a leaf, above the
    settle threshold — stop. Otherwise buy the next tier and ask again.

Two consequences worth being explicit about, because both are features:

* Most of a real camera roll settles at Tier.METADATA. A photo with a GPS fix and a
  date is `Places > United Kingdom > London` and `Timeline > 2025` for free. The
  expensive tiers exist for the minority that needs them, and `tier_used` reports
  how large that minority actually is — which is the number that decides whether
  first-run cataloguing is affordable.

* When the evidence supports "Screenshots" and nothing narrower, the engine returns
  `Screenshots`. It does not guess a leaf. A library full of confidently wrong leaves
  is worse than one that admits the depth it earned, because the user cannot tell the
  two apart until they go looking for something.
"""

from __future__ import annotations

from typing import Iterable, List, Optional

from . import rules, taxonomy
from .context import LibraryContext
from .signals import AssetSignals, Tier
from .verdict import Assignment, Classification, Evidence, REVIEW_FLOOR

# A leaf at or above this needs no further signal. Chosen so that a single strong
# unambiguous signal (a screenshot subtype, a GPS fix, a named face) settles, while
# anything resting on one hedged signal does not.
SETTLE_AT = 0.85


class Classifier:
    """Stateless per asset; the library context is shared and read-only."""

    def __init__(self, context: LibraryContext, budget: Tier = Tier.TEXT):
        # `budget` is what makes the ablation runnable: classify the same library at
        # Tier.METADATA to see what the cheap signals alone are worth, without
        # maintaining a second engine that would drift from this one.
        self.ctx = context
        self.budget = budget

    # -- public ------------------------------------------------------------
    def classify(self, a: AssetSignals) -> Classification:
        c = Classification(a.asset_id)
        c.tiers_spent.add(Tier.METADATA)

        self._timeline(a, c)
        self._metadata_root(a, c)

        # Screenshots are exempt from the visual tiers. Face clustering and scene
        # classification on a chat window produce noise at full price — the only
        # signal that tells you anything about a screenshot is its text.
        visual_worth_it = not (a.is_screenshot or a.is_screen_recording)

        if visual_worth_it and not self._settled(c) and self.budget >= Tier.VISUAL:
            c.tiers_spent.add(Tier.VISUAL)
            self._scenes(a, c)
        if visual_worth_it and not self._settled(c) and self.budget >= Tier.FACES:
            c.tiers_spent.add(Tier.FACES)
            self._faces(a, c)
        if self.budget >= Tier.TEXT and self._wants_text(a, c):
            c.tiers_spent.add(Tier.TEXT)
            self._text(a, c)

        # Places runs LAST among the content roots, and that ordering is load-bearing.
        # Run first, a GPS fix settled every photographed document as `Places > … >
        # London` at 0.91 confidence, the escalation policy saw a settled leaf, and the
        # OCR pass that would have recognised the passport never happened. Nine of nine
        # foreground documents were lost that way, silently, with a plausible answer in
        # their place. Location is context; it is not what the thing IS.
        self._places(a, c)
        self._travel(a, c)
        self._prune_implied_ancestors(c)
        self._cross_list(c)
        self._unfiled(a, c)
        return c

    def classify_all(self, assets: Iterable[AssetSignals]) -> List[Classification]:
        return [self.classify(a) for a in assets]

    # -- escalation policy -------------------------------------------------
    def _settled(self, c: Classification) -> bool:
        p = c.primary
        return bool(p and taxonomy.is_leaf(p.path) and p.confidence >= SETTLE_AT)

    def _wants_text(self, a: AssetSignals, c: Classification) -> bool:
        """Whether to READ text that exists. Whether to SPEND an OCR pass in the first
        place is the C-1 gate, and it lives upstream in the indexer where the pixels
        are — `OCRGate.swift`. Duplicating that decision here on metadata the engine
        cannot verify would be two gates disagreeing about one photo.

        Screenshots always read their text: the subtype settles the root, and only the
        text can ever open the leaf."""
        if not a.ocr_ran or not a.ocr_text:
            return False
        root = taxonomy.root_of(c.primary.path) if c.primary else None
        if root == "Screenshots":
            return True
        return not self._settled(c)

    # -- tiers -------------------------------------------------------------
    def _timeline(self, a: AssetSignals, c: Classification) -> None:
        """Every asset is reachable from Timeline. This is the multi-index premise:
        one original, many entries (§3), and the one entry that never fails.

        §3 asks for **Year → Month → Day → Moment**, not just a year. Only the year was
        indexed, which meant "the photos from that afternoon" had nothing to resolve
        against — a whole level of the tree the Constitution names as first-class.

        Only the deepest node survives `_prune_implied_ancestors`, which is correct
        and not a loss: the browse tree rolls counts up from the leaves, so a year still
        shows the right total, and holding the same asset at four depths would count it
        four times inside its own parent.

        Moment is a **capture session**, not a clock hour: a run of photographs taken
        close together. Hour boundaries cut bursts in half — six frames at 17:59 and
        two at 18:01 are one moment and would land in two — so the moment is derived
        from neighbours in `LibraryContext`, and is absent rather than guessed when
        this asset has none.
        """
        if a.created_at is None:
            c.notes.append("no capture date — this asset cannot be placed on the timeline")
            return
        when = a.created_at
        # Built without %-d / %e: those are glibc extensions and this project's own
        # machine is Windows, where they raise. A date format is not worth a
        # platform-specific crash.
        year = taxonomy.ensure_node(f"Timeline > {when.year}")
        c.add(Assignment(year, [Evidence(
            "created_at", Tier.METADATA, 0.98,
            f"taken on {when.day} {when:%B %Y}")]))

        month = taxonomy.ensure_node(f"{year} > {when:%m %B}")
        c.add(Assignment(month, [Evidence(
            "created_at", Tier.METADATA, 0.98, f"taken in {when:%B %Y}")]))

        day = taxonomy.ensure_node(f"{month} > {when:%d}")
        c.add(Assignment(day, [Evidence(
            "created_at", Tier.METADATA, 0.98,
            f"taken on {when.day} {when:%B %Y}")]))

        moment = self.ctx.moment_for(a.asset_id)
        if moment:
            node = taxonomy.ensure_node(f"{day} > {moment.label}")
            c.add(Assignment(node, [Evidence(
                "created_at", Tier.METADATA, 0.9,
                f"one of {moment.asset_count} photos taken around "
                f"{moment.start:%H:%M} that day")]))

    def _metadata_root(self, a: AssetSignals, c: Classification) -> None:
        if a.media_type == "video":
            # The canonical tree has eleven roots and none of them is Video. That is a
            # real gap (MNI-2 / FC-2), not something to paper over by filing video as
            # a photo, so it is recorded on the asset and surfaced by the catalogue.
            c.notes.append("video: the taxonomy has no node for video assets (FC-2)")

        if a.is_screenshot or a.is_screen_recording:
            c.add(Assignment("Screenshots", [Evidence(
                "is_screenshot", Tier.METADATA, 0.95,
                "the system recorded this as a screenshot")], is_primary=True))
            return

        if a.source == "downloaded":
            c.add(Assignment("Downloads", [Evidence(
                "source", Tier.METADATA, 0.88,
                "saved from another app rather than taken with the camera")], is_primary=True))
            return

    # Paperwork is not a place memory. A receipt photographed at the kitchen table
    # does not belong in `Places > United Kingdom > London` alongside the photos of
    # London — the shelf would stop meaning anything.
    _NOT_PLACES = frozenset({"Documents", "Purchases", "Screenshots", "Downloads"})

    def _places(self, a: AssetSignals, c: Classification) -> None:
        """A measured GPS fix is the single strongest cheap signal in a camera roll,
        and it is already in the metadata row. §11: an inferred fix is not this."""
        if a.geo is None or a.geo.source != "exif":
            return
        if a.place is None or not a.place.country:
            return
        if any(taxonomy.root_of(x.path) in self._NOT_PLACES for x in c.assignments):
            return

        parts = ["Places", a.place.country]
        reason = f"the location saved with the photo is in {a.place.country}"
        if a.place.city:
            parts.append(a.place.city)
            reason = f"the location saved with the photo is in {a.place.city}, {a.place.country}"
        path = taxonomy.ensure_node(taxonomy.SEP.join(parts))

        weight = min(0.92, 0.55 + 0.40 * a.place.confidence) * a.geo.confidence
        is_primary = not any(x.is_primary for x in c.assignments)
        c.add(Assignment(path, [Evidence("geo", Tier.METADATA, weight, reason)], is_primary=is_primary))

    def _travel(self, a: AssetSignals, c: Classification) -> None:
        """Travel is a library-level judgement, not a per-photo one: a run of days
        spent away from the home cluster. One photo in Tokyo is not a trip."""
        trip = self.ctx.trip_for(a.created_at, a.geo)
        if trip is None:
            return
        path = taxonomy.ensure_node(f"Travel > {trip.label}")
        span = (trip.end - trip.start).days + 1
        c.add(Assignment(path, [Evidence(
            "geo+created_at", Tier.METADATA, 0.80,
            f"part of {span} days away from home in {trip.city or trip.country or 'another place'}, "
            f"{trip.start.strftime('%B %Y')}")]))

    def _faces(self, a: AssetSignals, c: Classification) -> None:
        named = a.named_people
        if named:
            if len(named) == 1:
                path = taxonomy.ensure_node(f"People > {named[0]}")
                reason = f"{named[0]} was recognised in this photo"
                weight = 0.88
            else:
                path = "People > Groups"
                reason = "recognised " + ", ".join(named) + " together in this photo"
                weight = 0.85
            is_primary = not any(x.is_primary for x in c.assignments)
            c.add(Assignment(path, [Evidence("face_clusters", Tier.FACES, weight, reason)],
                             is_primary=is_primary))
        elif a.face_clusters:
            # An unnamed cluster still changes the risk class, which is the point:
            # "someone is in this photo" is enough to stop treating it as disposable.
            c.notes.append("unnamed person present")

    def _scenes(self, a: AssetSignals, c: Classification) -> None:
        best = {}
        for label in a.scene_labels:
            if label.confidence < rules.SCENE_FLOOR:
                continue
            path = rules.SCENE_MAP.get(label.identifier)
            if not path:
                continue
            if path not in best or label.confidence > best[path].confidence:
                best[path] = label
        for path, label in best.items():
            w = rules.scene_weight(label.confidence)
            pretty = label.identifier.replace("_", " ")
            is_primary = (not any(x.is_primary for x in c.assignments)
                          and taxonomy.root_of(path) in ("Objects", "Downloads"))
            c.add(Assignment(path, [Evidence(
                "scene_labels", Tier.VISUAL, w,
                f"a {pretty} was recognised in the picture")], is_primary=is_primary))

    def _text(self, a: AssetSignals, c: Classification) -> None:
        text = a.ocr_text or ""
        if not text.strip():
            return
        root = taxonomy.root_of(c.primary.path) if c.primary else None

        groups = [rules.SCREENSHOT_RULES, rules.DOCUMENT_RULES,
                  rules.PURCHASE_RULES, rules.WORK_TEXT_RULES]
        # Within a group the first match wins, so specific rules must precede general
        # ones in the table — `Documents > Identity > Passports` before `Documents`.
        for group in groups:
            for rule in group:
                if rule.requires_root and rule.requires_root != root:
                    continue
                if not rule.pattern.search(text):
                    continue
                is_primary = (not any(x.is_primary for x in c.assignments)
                              or (root == "Screenshots" and rule.requires_root == "Screenshots"))
                if root == "Screenshots" and rule.requires_root == "Screenshots":
                    # Deepen the existing Screenshots assignment rather than adding a
                    # competing root.
                    for existing in c.assignments:
                        if existing.path == "Screenshots":
                            existing.is_primary = False
                c.add(Assignment(rule.path, [Evidence("ocr_text", Tier.TEXT, rule.weight, rule.reason)],
                                 is_primary=is_primary))
                break

    def _prune_implied_ancestors(self, c: Classification) -> None:
        """`Screenshots` and `Screenshots > Temporary > Pickup Codes` are one entry,
        not two. The browse tree rolls counts up from the leaves, so leaving both in
        would count the asset twice in its own parent.

        The ancestor's evidence is merged into the descendant rather than discarded —
        "the system recorded this as a screenshot" is half the reason the leaf is
        right, and the user is owed the whole reason."""
        held = {x.path for x in c.assignments}
        for assignment in list(c.assignments):
            deeper = [p for p in held
                      if p != assignment.path and assignment.path in taxonomy.ancestors(p)]
            if not deeper:
                continue
            target = min(deeper, key=lambda p: taxonomy.depth(p))
            for other in c.assignments:
                if other.path == target:
                    other.evidence.extend(assignment.evidence)
                    other.is_primary = other.is_primary or assignment.is_primary
                    break
            c.assignments.remove(assignment)
            held.discard(assignment.path)

    def _cross_list(self, c: Classification) -> None:
        """`_also_in` from the canonical tree. One original, two entries (§3) — the
        receipt the user thinks of as paperwork and the receipt they think of as a
        purchase are the same photo, filed once."""
        for assignment in list(c.assignments):
            also = taxonomy.cross_listing(assignment.path)
            if not also:
                continue
            c.add(Assignment(also, [Evidence(
                "taxonomy", Tier.METADATA, min(0.9, assignment.confidence),
                f"also filed under {also} — the same photo, not a copy "
                f"(cross-listed from {assignment.path})")],
                cross_listed_from=assignment.path))

    def _unfiled(self, a: AssetSignals, c: Classification) -> None:
        """An asset with a date and nothing else is on the timeline and nowhere else.
        Saying so is more useful than inventing a home for it, and it is the honest
        count of how much of a library the engine cannot yet explain."""
        content = [x for x in c.assignments if taxonomy.root_of(x.path) != "Timeline"]
        if not content:
            c.notes.append("unfiled: no signal placed this asset anywhere but the timeline")
